### PR TITLE
feat(jwt): add support for JSON web tokens

### LINKED ISSUE #97: Create JWT decorator
**Is your feature request related to a problem? Please describe.**
JWT tokens provide an added layer of security to verify the integrity of the issuer of the request.

**Describe the solution you'd like**
For endpoints that require authentication, create a decorator to wrap the lambda function. The decorator should verify the integrity of the JWT token sent from the client. If the JWT token is invalid or expired, then the lambda function should return an HTTP 400 Bad Request.

**Describe alternatives you've considered**
None

**Additional context**
Possible sample code:

```python
def auth(fn)
    def wrapper(event, context):
        # if JWT token is invalid return 400
        try:
            jwt.decode(...)
        except Exception:
            return {'statusCode': 400}
        return fn(event, context)
    return wrapper

@auth
def handler(event, context):
    # do something
    return {
        'statusCode': 200
    }
```

### PR DESCRIPTION
**Pull request by jasonwalsh**
### Description

This pull request adds a decorator (`auth`) that does the following:

1. Decodes the JSON web token from the `Authorization` header
2. Performs a `get_item` operation on the DynamoDB table using the `merchant_id` from the payload of the decoded JWT token

See the flowchart in #71 for more details.

### Testing

```
python -m pytest tests/unit
```

Resolves #97

### CODE REVIEW

File: functions/decorators.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+from jwt.exceptions import ExpiredSignatureError
+
+from functools import wraps
+
+
+def auth(fn):
+
+    @wraps(fn)
+    def wrapper(event, context):
+        header = event['headers']['Authorization'].split(' ')[-1]
```
Comment (jasonwalsh) [inline functions/decorators.py]: Should check if the `Authorization` key returns `None`. If it does, then return an HTTP Bad Request.

Comment (jasonwalsh) [reply functions/decorators.py]:
```python
header = event['headers'].get('Authorization')
if header is None:
  return {
    'statusCode': 400
  }
```

File: functions/decorators.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+        try:
+            decoded = jwt.decode(header)
+        except ExpiredSignatureError:
```
Comment (jasonwalsh) [inline functions/decorators.py]: Should also catch other [exceptions](https://pyjwt.readthedocs.io/en/latest/api.html#exceptions).

File: functions/decorators.py (hunk 1 above)
Comment (brh55) [reply functions/decorators.py]: Agreed. Update if you can

File: functions/decorators.py (hunk 2 above)
Comment (brh55) [reply functions/decorators.py]: Anyway we can also provide a status message (expired, account error, etc)

File: functions/decorators.py (hunk 1 above)
Comment (brh55) [reply functions/decorators.py]: Also might be wise to add a test to make sure no tokens return a 400

Comment (jasonwalsh) [reply functions/decorators.py]: > Also might be wise to add a test to make sure no tokens return a 400

I am not sure what you mean

File: functions/decorators.py (hunk 2 above)
Comment (jasonwalsh) [reply functions/decorators.py]: Definitely. Would you prefer to the response payload to be something like:

```json
{
  "message": "<exception message>"
}
```

File: functions/decorators.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+def requires_auth(fn):
+
+    @wraps(fn)
+    def wrapper(event, context):
+        header = event['headers'].get('Authorization')
+
+        if header is None:
+            return {
+                'statusCode': HTTPStatus.UNAUTHORIZED
```
Comment (jasonwalsh) [inline functions/decorators.py]: I made the function return HTTP 401 Unauthorized if the `Authorization` header is not present.

File: functions/decorators.py (hunk 1 above)
Comment (jasonwalsh) [reply functions/decorators.py]: > Agreed. Update if you can

See a24f3a8

File: functions/decorators.py (hunk 3 above)
Comment (brh55) [reply functions/decorators.py]: 👍

**Pull request merged**